CARDS-2514: New locking module

acrowthe opened the pull request:
Implement the ability to lock and unlock subjects and forms as well as user groups for who is allowed to lock and unlock subjects and forms.
Adds:
- `cards:locked` restriction, allowing ACLs to be only apply to locked forms/subjects
- `Locks` resource homepage containing `cards:Lock` nodes specifying who locked something and at what time
- `cards:lock` property on forms and subjects containing a reference to a lock node when something is locked
- Lock/Unlock button on subjects and forms
- Locked status flag

Additional details:
- Children of a locked parent cannot be unlocked
- Unlocking a parent will unlock all locked children that were locked as a part of the original lock operation. Children that were locked prior to the parent being locked will not be unlocked

Testing:
- Verify that locking is opt-in:
  - Launch using `./start_cards.sh --test --dev` to make sure nothing has changed and that locking is disabled by default
  - Launch using `./start_cards.sh --test --dev --locking` to test that locking is can be enabled and is working as expected
  - Verify that locking can be enable using docker:
    `python3 generate_compose_yaml.py --cards_project cards4prems --oak_filesystem --dev_docker_image --locking`
    `docker-compose build && docker-compose up`
      - Test that locking is working as expected
    `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
  - Verify docker runs without locking by default:
    `python3 generate_compose_yaml.py --cards_project cards4prems --oak_filesystem --dev_docker_image`
  - Verify that locking works properly with `--permissions trusted` runmode

Make sure that:
- Only users in the LockUsers group can lock subjects
- Only users in the UnlockUsers group can unlock subjects
- Only admins can edit locked forms
- Subjects cannot be locked or unlocked if they have a locked parent
- If you lock a child subject then lock a parent subject, unlocking the parent will not unlock the child
- If you lock a parent subject that has unlocked children (forms and/or subject), the child nodes will be locked too

acrowthe committed:
CARDS-2514: New "locking" module
Add optional locking module

acrowthe committed:
CARDS-2514: New "locking" module
Add locking UI
Add subject action extension allowing for other modules to add actions to the subject page

acrowthe committed:
CARDS-2514: New "locking" module
Implement locking backend

acrowthe committed:
CARDS-2514: New "locking" module
Add component that adds locking permissions to trusted users on startup

acrowthe committed:
CARDS-2514: New "locking" module
Clean up debug logging

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -218,7 +229,15 @@ function SubjectContainer(props) {
-      <SubjectMember classes={classes} id={id} level={currentLevel} data={subject} maxDisplayed={maxDisplayed} pageSize={pageSize} onDelete={() => {setDeleted(true)}} childSubjects={relatedSubjects}/>
+      <SubjectMember
+        classes={classes}
+        id={id} level={currentLevel}
```
I'd put `level` on it's own line

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,325 @@
+//  under the License.
+//
+import React, { useState, useEffect, useContext } from "react";
+import { withRouter } from "react-router-dom";
```
Why do you need `withRouter` here?

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,325 @@
+  let handleIncompleteForms = (rows) => {
+    setNextAction(ACTION_CONTINUE);
+    setActionContent(getLockWarning());
+    setDialogHeader(<Typography variant="body1">{rows.length} form{rows.length > 1 ? "s are" : " is"} incomplete:</Typography>)
```
`variant="body1"` is a default for `Typography`, can be dropped

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  let closeError = () => {
+    setErrorOpen(false);
```
We can either use `errorMessage` as a both message holder and `errorOpen` switch or since it's better to clean error message on close

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,325 @@
+
+  return( <>
+    <ErrorDialog open={errorOpen} onClose={closeError}>
+      <Typography variant="body1">{errorMessage}</Typography>
```
`variant="body1"` same as above, can be dropped

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,325 @@
+          <Button variant="outlined" size="small" onClick={closeDialog}>Cancel</Button>
+          <Button
+            variant="contained"
+            size="small"
```
In light if CARDS-2651 we need to have all buttons same default size across all app

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,325 @@
+  variant: PropTypes.oneOf(["icon", "text", "extended"]), // "extended" means both icon and text
+}
+
+SubjectLockAction.defaultProps = {
```
Lets adopt the new React-19-friendly way of declaring defaults for props

veronikaslc commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockedRestrictionPattern.java:
```diff
@@ -0,0 +1,95 @@
+        return isForm(node) || isSubject(node);
+    }
+
+    protected static boolean isForm(final Tree node)
```
Since we already have duplicated `isForm(Tree node)` in `UnsubmittedFormsRestrictionPattern` I'd suggest move it to `FormUtils` and use across app

veronikaslc commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockedRestrictionPattern.java:
```diff
@@ -0,0 +1,95 @@
+        return false;
+    }
+
+    protected static boolean isSubjectOrForm(final Tree node)
```
To save computational time of getting type and checking if node has type I'd combine those 2 in 

if (node.hasProperty("jcr:primaryType")) {
    String type  = node.getProperty("jcr:primaryType").getValue(Type.STRING);
    return type.matches("cards:Form|cards:Subject");
}
return false;

veronikaslc commented on modules/locking/src/main/java/io/uhndata/cards/locking/internal/LockManagerImpl.java:
```diff
@@ -0,0 +1,512 @@
+        boolean mustCloseResolver = initializeServiceResolver();
+        try {
+            String reason = canLockWithReason(node, true);
+            return reason != null;
```
Looks like a logical error

veronikaslc commented on modules/locking/src/main/java/io/uhndata/cards/locking/internal/LockManagerImpl.java:
```diff
@@ -0,0 +1,512 @@
+        boolean mustCloseResolver = initializeServiceResolver();
+        try {
+            String reason = canUnlockWithReason(node);
+            return reason != null;
```
Here too

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -733,6 +737,20 @@ const questionnaireStyle = theme => ({
+      border: "1px solid " + theme.palette.action.disabled,
+      background: "transparent",
+      color: theme.palette.text.disabled,
+        zoom: .75,
```
slightly misallighned

veronikaslc commented on modules/locking/src/main/java/io/uhndata/cards/locking/internal/IncompleteFormsPrecondition.java:
```diff
@@ -0,0 +1,91 @@
+        try {
+            if (node.isNodeType(SUBJECT_NODE_TYPE)) {
+                checkSubject(node);
+            } else if (node.isNodeType(FORM_NODE_TYPE)) {
```
We have `isForm(node)` in `FormUtils` for it

veronikaslc commented on modules/locking/src/main/java/io/uhndata/cards/locking/internal/IncompleteFormsPrecondition.java:
```diff
@@ -0,0 +1,91 @@
+    public boolean canLock(Node node) throws LockWarning, LockError
+    {
+        try {
+            if (node.isNodeType(SUBJECT_NODE_TYPE)) {
```
We have `isSubject(node)` in `SubjectUtils` for it

veronikaslc commented on modules/locking/src/main/java/io/uhndata/cards/locking/internal/LockManagerImpl.java:
```diff
@@ -0,0 +1,512 @@
+        try {
+            VersionManager versionManager
+                = this.serviceResolver.adaptTo(Session.class).getWorkspace().getVersionManager();
+            Node serviceNode = getServiceNode(node);
```
Here and in `unlockNode()` we can get Session via `this.serviceResolver.adaptTo(Session.class)` once and pass down to `getServiceNode()` as a param so it can use it or get it's own if none is passed, just to save on operations.

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  let handleOpenDialogLocked = () => {
+    setDialogTitle(`Unlock ${entryType} "${subject?.identifier}"`)
```
some end of line `;` are missing

acrowthe committed:
CARDS-2514: New "locking" module
Addressing review comments (general code cleanup)

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectActions.jsx:
```diff
@@ -0,0 +1,61 @@
+  let { subject, reloadSubject, className, size, variant } = props;
+
+  const [ extensionPointComponents, setExtensionPointComponents ] = useState([]);
+  const [ isInitialized, setIsInitialized ] = useState(false);
```
No need to use a state variable here. The simple `let isInitialized= false;` and setting it to `true` inside the useEffect() without watching it is enough.
See the [reference](https://react.dev/learn/you-might-not-need-an-effect#initializing-the-application)

veronikaslc replied:
Same holds for other places of `setInitialized` used

acrowthe committed:
CARDS-2514: New "locking" module
Address review comments
- General code cleanup
- Split LockError into Error and Exception

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -420,7 +420,7 @@ function SubjectMemberInternal (props) {
       fetchTableData();
     }
-  }, [data]);
+  }, [data]['jcr:uuid']);
```
Typo, should be `[data['jcr:uuid']]`

marta- replied:
Addressed in 00aa19f .

marta- committed:
CARDS-2514: New "locking" module

Fixed typo that caused the Subject page to fetch data infinitely.

marta- commented:
Remaining issue: unlocking fails

![image](https://github.com/user-attachments/assets/865b3304-5dae-496a-9c5d-78c7730f6072)

![image](https://github.com/user-attachments/assets/3eb0ea5a-11cb-456c-9d8d-90efdca28ea2)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -733,6 +737,20 @@ const questionnaireStyle = theme => ({
         paddingTop: theme.spacing(4),
       },
     },
+    lockDialogInput: {
+      marginTop: theme.spacing(3),
+      marginRight: theme.spacing(3),
+      marginBottom: theme.spacing(3),
+    },
+    lockDialogAvatar: {
+      border: "1px solid " + theme.palette.action.disabled,
+      background: "transparent",
+      color: theme.palette.text.disabled,
+      zoom: .75,
+    },
```
These rules are only used by the locking module, so it's better to avoid declaring them in the data-entry module.

marta- replied:
Addressed in 92c0949 .

marta- commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,346 @@
+        {dialogContent}
+      </DialogContent>
+      <DialogActions className={classes.dialogActions}>
+          {actionContent}
+          <Button
```
The layout here  looks awkward, with the Alert being too close vertically to the buttons. It might be better to put it in a separate DialogActions.

marta- replied:
Addressed in 92c0949 .

marta- committed:
CARDS-2514: New "locking" module

Frontend tweaks:
* Moved the styling that was specific to the locking dialog from the data-entry module to inline styling
* Better layout of DialogActions

marta- commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,355 @@
+          }
+          return <ListItem key={index}>
+              <ListItemAvatar>
+                <Avatar sx={theme => ({
+                  border: "1px solid " + theme.palette.action.disabled,
+                  background: "transparent",
+                  color: theme.palette.text.disabled,
+                  zoom: .75
+                })}>
+                  <AssignmentIcon/>
+                </Avatar>
+              </ListItemAvatar>
```
We should consider creating a reusable `SurveyAvatar` component. We will soon have avatars styled like this in 3 places:
* patient portal
* clinician dashboard
* locking dialog

https://phenotips.atlassian.net/browse/CARDS-2667

marta- committed:
CARDS-2514: New "locking" module

Fix console warning caused by passing a 'string` value to the to the dialog's `open` prop which expects `boolean`

acrowthe committed:
CARDS-2514: New "locking" module
- Fix checkin error that was causing unlock to fail and being logged by locking
- Improve error messages displayed to the user when locking/unlocking fails

acrowthe committed:
CARDS-2514: New "locking" module
Improve a UX in a couple error cases:
- Silently do nothing when unlocking an unlocked node
- Display lock author and time when trying to lock a locked node
Fix button sizes
Some front end code cleanup

acrowthe committed:
CARDS-2514: New "locking" module
Move /Locks folder to /jcr:system/cards:locks
Fix error message typo

acrowthe committed:
CARDS-2514: New "locking" module
Remove locking webpack and package files due to #1940

sdumitriu committed:
CARDS-2514: New "locking" module
Bugfix: Production docker images don't include the locking module

marta- commented on modules/locking/src/main/java/io/uhndata/cards/locking/internal/LockManagerImpl.java:
```diff
@@ -0,0 +1,540 @@
+
+            final Set<String> statusFlags = getStatusFlags(serviceNode);
+            statusFlags.add(LOCKED_FLAG);
+            serviceNode.setProperty(STATUS_PROPERTY, statusFlags.toArray(new String[0]));
```
The setting of the status flag should be done in a commit editor. This code is not the only way to set locks, it can technically be done via APIs or using bin/browser. The status flag should be synced with the existence of the lock, regardless of how the lock gets created.

acrowthe committed:
CARDS-2514: New "locking" module
Move status flag updates into a editor to make sure the status flags reflect the true lock state

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -577,6 +577,10 @@ const questionnaireStyle = theme => ({
     },
+    LOCKEDFlag:{
+      backgroundColor: theme.palette.primary.main,
+      color: theme.palette.background.paper
```
Really inconsistent indentation in this file...

sdumitriu committed:
[misc] Update module version

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectActions.jsx:
```diff
@@ -0,0 +1,61 @@
+export default function SubjectActions(props) {
+  let { subject, reloadSubject, className, size, variant } = props;
+
+  const [ extensionPointComponents, setExtensionPointComponents ] = useState([]);
```
These actions are the same all the time for all subjects, aren't they? I would make this a static variable instead of a component state, since right now the subject actions are fetched for every displayed subject, which for a subject with many visits means lots of requests for the same information.

sdumitriu commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,357 @@
+        setDialogContent(
+          <>
+            <TextField
+              inputProps={{readonly:true}}
```
> Warning: Invalid DOM property `readonly`. Did you mean `readOnly`?

Both in Chrome and Firefox, and the input is indeed editable.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockedStatusFlagEditor.java:
```diff
@@ -0,0 +1,128 @@
+    // The property that contains a node's status flags
+    private static final String STATUS_FLAGS = "statusFlags";
+    // The property that contains a reference to a lock node if a node is locked
+    private static final String LOCK_PROPERTY = "cards:Lock";
```
Properties should start with a lowercase letter, so `cards:lock`. Also, the ACL that prevents random people from locking is set on `cards:lock`, so with the uppercase letter it is possible for non-lock users to apply locks by manually setting this property.

sdumitriu replied:
`cards:Lock` is hardcoded in many places actually, they should all refer to `LockManager.LOCK_PROPERTY`.

sdumitriu commented on modules/locking/src/main/resources/SLING-INF/nodetypes/locking.cnd:
```diff
@@ -0,0 +1,66 @@
+
+  // Children
+
+  // The list of forms.
```
```suggestion
  // The list of locks.
```

sdumitriu commented on modules/locking/src/main/features/feature.json:
```diff
@@ -0,0 +1,43 @@
+        // Locking and unlocking subjects and appropriate visits
+        "create service user locking \n set ACL for locking \n allow jcr:all on / \n end",
+        "set ACL for everyone \n deny rep:write on /Subjects,/Forms restriction(cards:locked) \n deny rep:write on /jcr:system/cards:locks \n allow jcr:read on /jcr:system/cards:locks \n deny rep:write on /Subjects,/Forms restriction(rep:itemNames, cards:lock) \n end",
+        "create group LockUsers \n set ACL for LockUsers \n allow jcr:addChildNodes, rep:addProperties, jcr:nodeTypeManagement on /jcr:system/cards:locks \n end",
+        "create group UnlockUsers \n set ACL for UnlockUsers \n allow jcr:removeChildNodes, jcr:removeNode on /jcr:system/cards:locks \n end"
```
I don't like this, any unlock user can `DELETE` any lock node without going through the unlock API, which leaves dangling lock references and LOCKED flags. It is then impossible to properly unlock the node, since the lock manager cannot find the lock node to delete.

sdumitriu replied:
I would make all changes in `/jcr:system/cards:locks/` using the `locking` service user, and simply check in the lock manager that the current user is a member of the right group.

sdumitriu commented on modules/locking/src/main/features/feature.json:
```diff
@@ -0,0 +1,43 @@
+        // Locking and unlocking subjects and appropriate visits
+        "create service user locking \n set ACL for locking \n allow jcr:all on / \n end",
+        "set ACL for everyone \n deny rep:write on /Subjects,/Forms restriction(cards:locked) \n deny rep:write on /jcr:system/cards:locks \n allow jcr:read on /jcr:system/cards:locks \n deny rep:write on /Subjects,/Forms restriction(rep:itemNames, cards:lock) \n end",
+        "create group LockUsers \n set ACL for LockUsers \n allow jcr:addChildNodes, rep:addProperties, jcr:nodeTypeManagement on /jcr:system/cards:locks \n end",
```
I don't like this, any lock user can create locks through HTTP requests. With the fixed `cards:lock` property name, they at least cannot add these locks to a subject, but still they should not be allowed to just create useless lock nodes.

sdumitriu commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,357 @@
+      handleOpenDialogLocked();
+    } else {
+      handleOpenDialogUnlocked();
+      fetchIncompleteForms();
```
This is also called in `handleOpenDialogUnlocked`, resulting in two requests being sent. Choose one place for the call.

sdumitriu reviewed (changes requested):
I forgot to mention: instead of putting all lock nodes under `/jcr:system/cards:locks`, I would create nested folders like we do with tokens. See how [`createParents`](https://github.com/data-team-uhn/cards/blob/7fbab95967e0f2c40800a42c77427ab7a3611e95/modules/token-authentication/src/main/java/io/uhndata/cards/auth/token/impl/TokenManagerImpl.java#L215) does it.

acrowthe committed:
CARDS-2514: New "locking" module
- Improve permissions
  - Instead of lock/unlock permissions being based on ability to write or delete nodes,
    it is now based on group membership and only the lock service user can edit locks
- Clean up code style and node naming
- Reduce SubjectAction redundant fetches
- Move locks into folders similar to tokens

acrowthe committed:
CARDS-2514: New "locking" module
- Fix bug with SubjectActions where the first SubjectActions to render may never re-render
  once it has finished fetching the list of actions
- Improve lock/unlock button to not be visible if the subject's parent is locked

acrowthe commented:
Fixed a bug where the first SubjectAction may never re-render after loading the list of actions/buttons and improved the lock/unlock button to not be visible if the parent is locked

marta- commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,366 @@
+  }
+
+  let getLockWarning = () => {
+    return <Alert severity="warning">Once you sign off on this {entryType}, it will be locked along with
+            all the associated data forms and no further edits will be possible. Proceed?</Alert>
```
Slightly easier to read:
```suggestion
    return (
      <Alert severity="warning">
        Once you sign off on this {entryType}, it will be locked along with
        all the associated data forms and no further edits will be possible. Proceed?
      </Alert>
    );
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -120,7 +122,15 @@ function Subject(props) {
+          getSubject={handleSubject}
+          reloadSubject={fetchRelatedRef}
+          history={history}
+          contentOffset={props.contentOffset}/>
```
Very minor code style suggestion:
```suggestion
          contentOffset={props.contentOffset}
          />
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -218,7 +229,16 @@ function SubjectContainer(props) {
+        pageSize={pageSize}
+        onDelete={() => {setDeleted(true)}}
+        childSubjects={relatedSubjects}
+        fetchSubjectData={fetchSubjectData}/>
```
Very minor code style suggestion:
```suggestion
        fetchSubjectData={fetchSubjectData}
        />
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -616,7 +651,15 @@ function SubjectMemberInternal (props) {
+                maxDisplayed={maxDisplayed}
+                pageSize={pageSize}
+                subject={subject}
+                fetchSubjectData={fetchSubjectData}/>
```
```suggestion
                fetchSubjectData={fetchSubjectData}
                />
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectActionContext.jsx:
```diff
@@ -0,0 +1,27 @@
+    value: []
+};
+
+export default SubjectActionContext;
```
```suggestion
export default SubjectActionContext;

```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectActions.jsx:
```diff
@@ -0,0 +1,68 @@
+  // if the list of actions from SubjectActionContext is used directly for rendering.
+  let [ actions, setActions ] = useState([]);
+
+  const FETCHING = "Fetching"
+  const LOADED = "Loaded"
```
I would name these STATUS_FETCHING and STATUS_LOADED.

marta- reviewed (commented):
First pass, just a couple of minor code style comments.

acrowthe committed:
CARDS-2514: New "locking" module
Frontend code style improvements

marta- commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -324,7 +331,8 @@ function SubjectLockAction(props) {
       </DialogActions>
     </Dialog>
-    { variant == "icon" ?
+    { displayButton ? (
```
We shouldn't be rendering the Dialogs either if the button doesn't need get displayed.

acrowthe committed:
CARDS-2514: New "locking" module
Adjust "hide button when can't lock/unlock" to hide entire action

veronikaslc committed:
CARDS-2514: New locking module
Refactor TextField input props

veronikaslc committed:
CARDS-2514: New locking module
More consistent whitespacing

veronikaslc commented:
Sometimes we provide information of how many forms are associated with the subject. Would it be useful here?
![image](https://github.com/user-attachments/assets/62592eb1-9a41-4436-bcd3-10c657a42dc4)

veronikaslc commented:
I made a build, created patient and tried to lock it logged as an admin-> got 500 respond with NPE
No errors showed on front end, also can not log in back via login dialog
![image](https://github.com/user-attachments/assets/10a42c04-2285-4198-aab6-628b38f76117)
![image](https://github.com/user-attachments/assets/91ef532a-3511-41d4-b4c0-d4034d8ba298)
![image](https://github.com/user-attachments/assets/95d62380-e554-4c72-b803-24fb209e8d9c)

veronikaslc commented:
> Had admin been added to the LockUsers group?

There are no groups to add to on start. If they have to be manually created that might be in improving point for instructions or autocreation.

acrowthe commented:
> There are no groups to add to on start. If they have to be manually created that might be in improving point for instructions or autocreation.

Definitely an issue then, there should be a LockUsers and UnlockUsers group that is auto-created

marta- commented:
> Sometimes we provide information of how many forms are associated with the subject. Would it be useful here?

Good idea. We list the _incomplete_ forms if any, but it might be useful to inform the user how many visits and how many forms we're locking with this patient. We just need to find a user friendly (not too technical) way to present it the information.

> Should "Lock" word be capitalized?

"Lock" should not be capitalized.

sdumitriu committed:
CARDS-2514: New locking module
Bugfix: since the order of the individual repoinit fragments is random, make sure /Forms and /Subjects exist before trying to apply restrictions on them

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockedStatusFlagEditor.java:
```diff
@@ -0,0 +1,127 @@
+        throws CommitFailedException
+    {
+        if (this.isLockableNode) {
+            final Set<String> statusFlags = new HashSet<>();
```
This randomizes the flags, use a `LinkedHashSet` instead.

acrowthe committed:
CARDS-2514: New locking module
- Fix lock header capitalization
- Change status editor set type to avoid accidental reordering

sdumitriu committed:
[bugfix] Since the order of the individual repoinit fragments is random, make sure /Forms and /Subjects exist before trying to apply restrictions on them

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,373 @@
+import ErrorDialog from "../components/ErrorDialog.jsx";
+
+function SubjectLockAction(props) {
+  const {
```
```
PropTypes.checkPropTypes(SubjectLockAction.propTypes, props, 'prop', 'SubjectLockAction');
```

sdumitriu committed:
[misc] Refactor configurable actions loading mechanism

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockServlet.java:
```diff
@@ -0,0 +1,157 @@
+    @Override
+    protected StringBuffer getAllowedRequestMethods(Map<String, Method> declaredMethods)
+    {
+        StringBuffer allowBuf = new StringBuffer();
```
The base javadoc states that overrides should always call the super method.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockServlet.java:
```diff
@@ -0,0 +1,157 @@
+        } catch (AccessDeniedException e) {
+            writeError(response, SlingHttpServletResponse.SC_FORBIDDEN, "Access Denied");
+        } catch (RepositoryException e) {
+            LOGGER.error("Unable to write response", e);
```
This doesn't look like the right message.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/api/LockManager.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @param node the subject node to unlock
+     * @throws LockError if the node cannot be unlocked
+     * @throws LockExcepttion if an internal error occurs
```
```suggestion
     * @throws LockException if an internal error occurs
```

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/internal/LockRepositoryInitializer.java:
```diff
@@ -0,0 +1,112 @@
+                    "/Forms",
+                    false,
+                    trustedUser.getPrincipal(),
+                    new String[] { repWrite },
```
`repWrite` could already be a string array instead of having to create a new string array four times.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/api/LockManager.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @param node the subject node to try to lock
+     * @throws LockError if the node can not be locked
+     * @throws LockExcpetion if an internal error occurs
```
```suggestion
     * @throws LockException if an internal error occurs
```

sdumitriu commented on modules/locking/src/main/frontend/assets.config:
```diff
@@ -0,0 +1,3 @@
+// These lines are aggregated as module.exports entries in webpack.config.js file by the webpack_script.py in aggregated-frontend module
+
+['cards-locking.SubjectLockAction']: './src/locking/SubjectLockAction.jsx'
```
This needs its dependencies declared, or else in production mode it will crash.

sdumitriu committed:
CARDS-2514: New locking module
Declare the dependencies of the SubjectLockAction component

sdumitriu committed:
CARDS-2514: New locking module
Check prop types using the new mechanism

sdumitriu committed:
CARDS-2514: New locking module
Apply codestyle, fix some typos, update @since version

sdumitriu committed:
CARDS-2514: New locking module
Append instead of replace list of allowed HTTP methods

veronikaslc commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -0,0 +1,375 @@
+      setRequestInProgress(false);
+      if (response.ok)  {
+        closeDialog();
+        reloadSubject && reloadSubject();
```
```suggestion
        reloadSubject?.();
```

acrowthe committed:
CARDS-2514: New locking module
Fix bug where locking data reload didn't correctly update target when navigating
subjects by breadcrumb

sdumitriu committed:
[misc] Use optional chaining for shorter code

Co-authored-by: veronikaslc <veronika.koltunov@gmail.com>
